### PR TITLE
Adding fixusings support

### PR DESCRIPTION
**Pull request by sriramgd**
1. Add missing usings
2. Send a dictionary of key-list for ambiguous usings
3. Remove redundant usings

TODO: 
1. Remove unused using statements
2. Sort using statements
3. If possible, replace RemoveRedundantUsings by something better from the framework

### CODE REVIEW

**Comment by Yantrio** [comment]
can we switch the bracing from k+r? apart from that, it looks good in theory to me.

**Comment by mat-mcloughlin** [comment]
Is it possible to make the response to match the old server so that we don't break backwards compatibility? 

https://github.com/OmniSharp/omnisharp-server/blob/master/OmniSharp/CodeActions/FixUsingsResponse.cs

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            var compilationUnitSyntax = (CompilationUnitSyntax)(root);
+            var usingsToAdd = new HashSet<string>();
+            response.UnresolvedDict = new Dictionary<string, List<string>>();
+
+            if (compilationUnitSyntax != null) {
+                foreach (var unresolvedType in unresolvedTypes) {
+                    var candidateUsings = SymbolFinder.FindDeclarationsAsync(document.Project, 
+                            unresolvedType.Identifier.ValueText, false, SymbolFilter.All, CancellationToken.None).Result;
+
+                    var isAmbiguousSymbol = (candidateUsings.Count() > 1);
```
Comment (Yantrio) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: this isn't needed, can be done inline

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                    foreach (var candidateUsing in candidateUsings) {
+                        var candidateNameSpace = GetFullNameSpace(candidateUsing, ignoreName: true);
+                        var candidateName = SyntaxFactory.ParseName(candidateNameSpace);
+                        var usingToAdd = SyntaxFactory.UsingDirective(candidateName).NormalizeWhitespace()
+                            .WithTrailingTrivia(SyntaxFactory.Whitespace("\n"));
+                        var usingToAddText = usingToAdd.GetText().ToString();
+                        if (!isAmbiguousSymbol) {
```
Comment (Yantrio) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: candidateUsings.Any() could do better here

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+using Microsoft.CodeAnalysis.Simplification;
+using OmniSharp.Models;
+
+namespace OmniSharp
+{
+    public partial class OmnisharpController
+    {
+        //TODO: 
+        //1. Remove unused using statements
+        //2. Sort using statements
```
Comment (Yantrio) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: does this todo list need updating/removing?

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs (hunk 2 above)
Comment (nosami) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: Don't think so... needs to check for > 1, not > 0

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                            }
+                        }
+                    }
+                }
+            }
+
+            return compilationUnitSyntax;
+        }
+
+        private async Task<CompilationUnitSyntax> RemoveRedundantUsings(CompilationUnitSyntax compilationUnitSyntax,
```
Comment (nosami) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: This is just removing dupes?

File: tests/OmniSharp.Tests/FixUsingsFacts.cs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+        private async Task AssertUnresolvedReferences(string fileContents, Dictionary<string, List<string>> expectedUnresolvedDict) 
+        {
+            var response = await RunFixUsings(fileContents);
+            foreach (var k in response.UnresolvedDict.Keys) {
+                var count = 0;
+                foreach (var x in response.UnresolvedDict[k]) {
+                    Assert.Contains(k, expectedUnresolvedDict.Keys);
+                    Assert.Equal(x, (expectedUnresolvedDict[k])[count]);
+                    //System.Console.WriteLine("Checked " + k + "->" + x);
```
Comment (nosami) [inline tests/OmniSharp.Tests/FixUsingsFacts.cs]: delete

**Comment by nosami** [comment]
Very happy to see this PR. 

It absolutely must use the [same response](https://github.com/OmniSharp/omnisharp-server/blob/master/OmniSharp/CodeActions/FixUsingsResponse.cs) as omnisharp v1.

Also... not sure there's much benefit from having remove usings if it's only removing dupes. We should probably delete that code and address that in a separate PR.

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+            {
+                var semanticModel = await document.GetSemanticModelAsync();
+                var compilationUnitSyntax = await AddMissingUsings(document, semanticModel, response);
+                await RemoveRedundantUsings(compilationUnitSyntax, response);
+            }
+
+            return response;
+        }
+
+        private async Task<CompilationUnitSyntax> AddMissingUsings(Document document,
```
Comment (davidfowl) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: Logic like this is usually moved into the https://github.com/OmniSharp/omnisharp-roslyn/tree/master/src/OmniSharp/Workers folder. The controllers are a thin layer that expose an end point calling into those workers.

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                await RemoveRedundantUsings(compilationUnitSyntax, response);
+            }
+
+            return response;
+        }
+
+        private async Task<CompilationUnitSyntax> AddMissingUsings(Document document, 
+                SemanticModel semanticModel, FixUsingsResponse response) 
+        {
+            var root = (await document.GetSyntaxTreeAsync()).GetRoot();
```
Comment (davidfowl) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: This code needs some comments

File: src/OmniSharp/Models/FixUsingsResponse.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+
+namespace OmniSharp.Models
+{
+    public class FixUsingsResponse
+    {
+        public string FileContents { get; set; }
+        public Dictionary<string, List<string>> UnresolvedDict {get; set;}
```
Comment (davidfowl) [inline src/OmniSharp/Models/FixUsingsResponse.cs]: Is that really the property name? Dict?

**Comment by davidfowl** [comment]
This code needs to be formatted using the format function in omnisharp :smile:

**Comment by nosami** [comment]
@sriramgd Are you planning on working on this some more? If not, I can pick it up where you left off.

**Comment by sriramgd** [comment]
@nosami I am. Will get in the updated PR by tomorrow evening. (Sorry about the break, was on vacation).

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs (hunk 1 above)
Comment (sriramgd) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: Done

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs (hunk 4 above)
Comment (sriramgd) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: Yes, removed.

File: tests/OmniSharp.Tests/FixUsingsFacts.cs (hunk 5 above)
Comment (sriramgd) [inline tests/OmniSharp.Tests/FixUsingsFacts.cs]: Deleted.

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs (hunk 6 above)
Comment (sriramgd) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: Moved.

File: src/OmniSharp/Models/FixUsingsResponse.cs (hunk 8 above)
Comment (sriramgd) [inline src/OmniSharp/Models/FixUsingsResponse.cs]: Removed and replaced with response matching v1 server.

**Comment by sriramgd** [comment]
Updated, ready for second review. Seems the travis build fails due to a previous merge.

**Comment by davidfowl** [comment]
Gonna fix the build in a sec

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Mvc;
+using OmniSharp.Models;
+
+namespace OmniSharp
+{
+    public partial class OmnisharpController
+    {
+        //TODO:
```
Comment (davidfowl) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: Move these todos, to the worker.

File: src/OmniSharp/Workers/Refactoring/FixUsings.cs [hunk 10]
```diff
@@ -0,0 +1,113 @@
+/* using Microsoft.CodeAnalysis.CSharp;
```
Comment (davidfowl) [inline src/OmniSharp/Workers/Refactoring/FixUsings.cs]: Dead code?

File: src/OmniSharp/Workers/Refactoring/FixUsings.cs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+{
+    public class FixUsingsWorker
+    {
+        public static async Task<FixUsingsResponse> AddMissingUsings(string fileName, Document document, SemanticModel semanticModel)
+        {
+            var root = (await document.GetSyntaxTreeAsync()).GetRoot();
+            var unresolvedTypes = root.DescendantNodes()
+                .OfType<SimpleNameSyntax>()
+                .Where(x => semanticModel.GetSymbolInfo(x).Symbol == null);
+            var compilationUnitSyntax = (CompilationUnitSyntax)(root);
```
Comment (davidfowl) [inline src/OmniSharp/Workers/Refactoring/FixUsings.cs]: `(CompilationUnitSyntax)root;` Get rid of the extra parenthesis

**Comment by nosami** [comment]
I just tried this out. It seems to almost be working, but there are quite a few issues with it I think.

Removing `using OmniSharp.Models;` from a controller and then running fix usings results in 2 usings being added back

```
using OmniSharp.Models;
using OmniSharp;
```
There are also lots of ambiguous messages when there should be none (see screenshot)
screen shot 2015-04-17 at 13 54 20

Lots of namespaces in the OmniSharp project don't get imported, e.g. Xunit. I tried this on lots of different classes. I've also seen it import namespaces that are already included.

**Comment by sriramgd** [comment]
Yes, this implementation is broken. Too many false positives for the symbols. I had only tested against the unit tests which were quite trivial. Will make better tests and look at fixing this.

**Comment by sriramgd** [comment]
I am blocked on this for now. The solution I have requires being able to update the Document with a CompilationUnitSyntax before doing the ambiguity check. However it seems like document.UpdateSyntaxRoot is not available in Roslyn now (looks like it used to be a while back). I have asked a question about this on SO. In the meantime, anyone else who knows better or has other ideas could take this forward.

**Comment by sriramgd** [comment]
Actually I did get a response on SO. Was missing something obvious. Will update PR in sometime.

File: src/OmniSharp/Workers/Refactoring/FixUsings.cs (hunk 11 above)
Comment (sriramgd) [inline src/OmniSharp/Workers/Refactoring/FixUsings.cs]: Done

File: src/OmniSharp/Workers/Refactoring/FixUsings.cs (hunk 10 above)
Comment (sriramgd) [inline src/OmniSharp/Workers/Refactoring/FixUsings.cs]: Removed

File: src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs (hunk 9 above)
Comment (sriramgd) [inline src/OmniSharp/Api/Refactoring/OmnisharpController.FixUsings.cs]: Moved

**Comment by sriramgd** [comment]
I have updated the implementation. This should do a better job at
1. Not including existing usings
2. Not flagging non-ambiguous symbols as ambiguous.

If this is accepted, we should replace any code which can be replaced by relevant APIs which Roslyn exposes in the future.

**Comment by nosami** [comment]
Just had a chance to try this again. It's definitely better, but there are still some bugs.

```cs
public class MyClass
{
    public MyClass()
    {
        var l = new List<string>();
    }
}
```

This code errored when trying to import usings.


Also..  does this code work with linq syntax? i.e. `from s in list where.....`. I couldn't test because of the previous error.

**Comment by sriramgd** [comment]
Sorry, had been busy for a while.
Seems like my implementation does not work well. Please close my PR in case anyone else wants to take this on.

**Comment by nosami** [comment]
Closing in favour of #291

**Pull request closed without merging**